File: scala/minimum_add_to_make_parentheses_valid.scala
//            Scala
//     Time complexity - O (N)
//     Space complexity - O (1)

object MinimumAddToMakeParenthesesValidScala {
  def minAddToMakeValid(s: String): Int = {
    var openParentheses = 0
    var minAdd = 0

    for (c <- s) {
      if (c == '(') {
        openParentheses += 1
      } else if (c == ')') {
        if (openParentheses == 0) {
          minAdd += 1
        } else {
          openParentheses -= 1
        }
      }
    }

    minAdd + openParentheses
  }
}

File: js/minimum_add_to_make_parentheses_valid.js
//             JS
//     Time complexity - O (N)
//     Space complexity - O (1)

var minAddToMakeValid = function minAddToMakeValid(s) {
    let openParentheses = 0;
    let minAdd = 0;

    for (let char of s) {
        if (char === '(') {
            openParentheses++;
        } else if (char === ')') {
            if (openParentheses === 0) {
                minAdd++;
            } else {
                openParentheses--;
            }
        }
    }

    return minAdd + openParentheses;
};